src/isar/storage/uploader.py

```diff
@@ -22,21 +70,66 @@ def run(self) -> None:
             try:
-                inspection, mission_metadata = self.upload_queue.get(timeout=5)
+                inspection, mission_metadata = self.upload_queue.get(timeout=1)
             except Empty:
+                if self.retry_queue:
+                    self._process_retry_queue()
                 continue
 
             for storage_handler in self.storage_handlers:
-                try:
-                    storage_handler.store(
-                        inspection=inspection, metadata=mission_metadata
-                    )
-                    self.logger.info(
-                        f"Storage handler: {type(storage_handler).__name__} "
-                        f"uploaded inspection {str(inspection.id)[:8]}"
-                    )
-                except StorageException:
-                    self.logger.warning(
-                        f"Storage handler: {type(storage_handler).__name__} "
-                        f"failed to upload inspection: {str(inspection.id)[:8]}"
-                    )
+                self._upload(
+                    storage_handler, inspection, mission_metadata, retry_count=-1
+                )
+
+    def _upload(
+        self,
+        storage_handler: StorageInterface,
+        inspection: Inspection,
+        mission_metadata: MissionMetadata,
+        retry_count: int,
+    ):
+        try:
+            storage_handler.store(inspection=inspection, metadata=mission_metadata)
+            self.logger.info(
+                f"Storage handler: {type(storage_handler).__name__} "
+                f"uploaded inspection {str(inspection.id)[:8]}"
+            )
+        except StorageException:
+            if retry_count < self.max_retry_attempts:
+                retry_item: UploaderRetryItem = UploaderRetryItem(
+                    inspection,
+                    mission_metadata,
+                    storage_handler,
+                    __retry_count=retry_count,
+                )
+                retry_item.increment_retry(self.max_wait_time)
+                self.retry_queue.append(retry_item)
+                self.logger.warning(
+                    f"Storage handler: {type(storage_handler).__name__} "
+                    f"failed to upload inspection: {str(inspection.id)[:8]}. "
+                    f"Retrying in {max(0, (retry_item.get_retry_time() - datetime.utcnow()).total_seconds())}s."
+                )
+            else:
+                self.logger.error(
+                    f"Storage handler: {type(storage_handler).__name__} "
+                    f"exceeded max retries to upload inspection: {str(inspection.id)[:8]}. Aborting upload."
+                )
+
+    @staticmethod
+    def _retry_is_ready(item: UploaderRetryItem) -> bool:
+        return datetime.utcnow() >= item.get_retry_time()
+
+    def _process_retry_queue(self):
+        ready_items: List[UploaderRetryItem] = list(
+            filter(lambda x: Uploader._retry_is_ready(x), self.retry_queue)
+        )
```

Comment:
```suggestion
        ready_items: List[UploaderRetryItem] = [x if Uploader._retry_is_ready(x) for x in self.retry_queue]
```
I'm not certain what is considered the best approach when it comes to using the filter function here, but this is an alternative in Python which is called a list comprehension. It should yield the same result I think and you don't have to cast to list etc.